Support Platform Network Reconciliation - Day-1 Workflows

sriram-gn opened the pull request:
This commit enables all of Day-1 workflows related to platform network reconciliation.
Apart from that, there are certain mechanisms built in to handle notifying host controller whenever an instance of platform network resource and address pool resource is created or updated in a non-blocking way while taking care of potential issues due to concurrent operations.

Test Cases:
 1. PASS - Verify successful unlock of AIO-SX in Day-1 without platformnetwork / addresspool instances.
 2. PASS - Verify successful Day-2 operation on AIO-SX host without platformnetwork / addresspool instances.
 3. PASS - Verify reconfiguration is blocked for all networks in Day-1.
 4. PASS - Verify fresh configuration (even partial ie. single new stack) is successful for all networks other than oam / admin / mgmt.
 5. PASS - Verify fresh configuration (even partial) is blocked for networks such as oam / admin / mgmt.
 6. PASS - Verify IPv6 addresspool associated with pxeboot fails to reconcile without requeuing the request despite fresh configuration.
 7. PASS - Verify oam fails to reconcile if either 'floatingAddress' or 'gateway' is missing from the addresspool spec on AIO-SX.
 8. PASS - Verify mgmt / cluster-host / pxeboot fails to reconcile if any of 'floatingAddress', controller0Address' or 'controller1Address' are missing from the addresspool spec.
 9. PASS - Verify addresspool resource instances can be deleted with no impact on the system.
10. PASS - Verify platformnetwork resource instances can be deleted with no impact on the system.

yjian118 commented on controllers/addresspool_controller.go:
```diff
@@ -13,7 +13,8 @@ import (
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/types"
+	// "k8s.io/apimachinery/pkg/types"
```
remove the package if no longer needed

yjian118 commented on controllers/manager/manager.go:
```diff
@@ -593,6 +606,60 @@ func (m *PlatformManager) CancelMonitor(object client.Object) {
 	}
 }
 
+func (m *PlatformManager) GetActiveHostAndSystemInfo(namespace string, client *gophercloud.ServiceClient) (*v1.Host, *SystemInfo, error) {
```
Suggest to split it into two methods: getActiveHost and getSystemInfo

sriram-gn commented on controllers/host/host_controller.go:
```diff
@@ -1486,29 +1482,29 @@ func (r *HostReconciler) StopAfterInSync() bool {
 // ReconcileExistingHost is responsible for dealing with the provisioning of an
 // existing host.
-func (r *HostReconciler) ReconcileExistingHost(client *gophercloud.ServiceClient, instance *starlingxv1.Host, profile *starlingxv1.HostProfileSpec, host *hosts.Host) error {
+func (r *HostReconciler) ReconcileExistingHost(client *gophercloud.ServiceClient, instance *starlingxv1.Host, profile *starlingxv1.HostProfileSpec, host *hosts.Host, request_namespace string) (error, []error) {
```
That's a good point but I thought of having platformnetwork subreconciler's error captured separately so that it will be easier to implement specific exception handling for subreconciler's errors. Unlike other sub-reconcilers, platformnetwork sub-reconciler is not directly involved in host attributes reconciliation so there might be instances where we just want to let certain errors not block host reconciliation. What do you think @yjian118 ?

yjian118 replied:
Not quite sure what do you mean "not block host reconciliation", as "inSync = (err == nil && platform_network_subreconciler_errs == nil)" is should be blocking the host status. Would suggest have clear message regarding the reason why it blocks. And you can define the type/struct of an error and guide the error handling by its type

yjian118 commented on controllers/host/host_controller.go:
```diff
@@ -1590,16 +1586,47 @@ func (r *HostReconciler) ReconcileExistingHost(client *gophercloud.ServiceClient
+			}
+
+		} else {
+			err_msg := "platform network reconciliation is already in progress in active controller."
```
As it is expected to only update in the active controller, can we change the message to "waiting for platform network reconciled"?

yjian118 commented on controllers/platformnetwork_controller.go:
```diff
@@ -385,7 +391,6 @@ func (r *PlatformNetworkReconciler) RestorePlatformNetworkStatus(instance *starl
 			instance.Status.Reconciled = true
 			instance.Status.ObservedGeneration = instance.ObjectMeta.Generation
 			instance.Status.DeploymentScope = "bootstrap"
-			instance.Status.StrategyRequired = "not_required"
```
Why to remove the not_required here?

sriram-gn replied:
The "StrategyRequired" field is not part of PlatformNetworkStatus anymore.
It is also not part of AddressPoolStatus.
The reason is VIM checks "strategy_required" of Host resource instance only to propagate VIM strategies (ie. lock/unlock) and the field seemed to serve no purpose on PlatformNetwork / AddressPool resources.

yjian118 replied:
acked

hugonicodemos commented on common/utilities.go:
```diff
@@ -163,3 +163,10 @@ func DedupeSlice[T comparable](sliceList []T) []T {
 }
 */
+
+func NormalizeIPv6(address string) string {
```
Can you handle the fail case?

    ip := net.ParseIP(address)
    if ip == nil {
       HANDLE FAIL
    }
    return ip.String()

sriram-gn replied:
Done

hugonicodemos commented on controllers/addresspool_controller.go:
```diff
@@ -62,20 +62,35 @@ func (r *AddressPoolReconciler) UpdateInsyncStatus(client *gophercloud.ServiceCl
+				msg := fmt.Sprintf("Failed to update '%s' host instance status", host_instance.Name)
+				logAddressPool.Error(err, msg)
+				return common.NewResourceConfigurationDependency(msg)
+			} else {
```
dont need the else (return inside the if), please avoid to turn the methods more nested.

sriram-gn replied:
Done

hugonicodemos commented on controllers/addresspool_controller.go:
```diff
@@ -62,20 +62,35 @@ func (r *AddressPoolReconciler) UpdateInsyncStatus(client *gophercloud.ServiceCl
+				msg := fmt.Sprintf("Failed to update '%s' host instance status", host_instance.Name)
+				logAddressPool.Error(err, msg)
+				return common.NewResourceConfigurationDependency(msg)
+			} else {
```
dont need the else (return inside the if), please avoid to turn the methods more nested.

sriram-gn replied:
Done

hugonicodemos commented on controllers/host/host_controller.go:
```diff
@@ -1590,9 +1586,45 @@ func (r *HostReconciler) ReconcileExistingHost(client *gophercloud.ServiceClient
+	}
+
+	if instance.Name == host_instance.Name {
+		// Only reconcile platform networks in active controller reconciliation
```
Can we modularize this in a utility method?

sriram-gn replied:
Done, modularized under `IsActiveHost` method in host controller.

hugonicodemos commented on controllers/host/platformnetworks.go:
```diff
@@ -5,72 +5,953 @@ package host
+// address and the second element is the range end address.  This is to align
+// with the system API formatting which represents a pair as an array of two
+// elements.
+func makeRangeArray(ranges []starlingxv1.AllocationRange) [][]string {
```
these 2 method should be moved to a common utils file.

sriram-gn replied:
Tried to move `makeRangeArray` and `compareRangeArrays` but then realized it causes circular import issue between `package v1` and `package common`, ie. `v1` imports `common` and to get `makeRangeArray` to work we have to import `v1` in `common` :)

hugonicodemos commented on controllers/host/platformnetworks.go:
```diff
@@ -5,72 +5,953 @@ package host
+		found_addrpool = GetSystemAddrPoolByUUID(addrpool_list, *addrpool_instance.Status.ID)
+		if found_addrpool != nil {
+			return found_addrpool, nil
+		} else {
```
else not needed

sriram-gn replied:
Done

hugonicodemos reviewed (changes requested):
Great job, @sriram-gn . Just few nits inline, and I recommend to move all these get by name/uuid functions to a common place, this could be necessary outside the platformnetwork controller and makes the file cleaner. thanks!

sriram-gn commented:
> Great job, @sriram-gn . Just few nits inline, and I recommend to move all these get by name/uuid functions to a common place, this could be necessary outside the platformnetwork controller and makes the file cleaner. thanks!

Thanks for reviewing @hugonicodemos!
I've moved name / uuid functions under utilities which definitely tidied platformnetworks.go a whole lot!

Also updated failure scenario in `NormalizeIPv6` and avoided `return` nested under `else` wherever I could.

It seems like we cannot move makeRangeArray under utilites and keeping compareRangeArrays following weird issue:
```
# github.com/wind-river/cloud-platform-deployment-manager/controllers/host
controllers/host/platformnetworks.go:256:39: undefined: utils.compareRangeArrays
# github.com/wind-river/cloud-platform-deployment-manager/controllers/host
vet: controllers/host/platformnetworks.go:256:39: undefined: utils.compareRangeArrays
```

So, I had to just let the two functions stick around in platformnetworks.go.

hugonicodemos commented on controllers/host/platformnetworks.go:
```diff
@@ -5,72 +5,911 @@ package host
+		// In case addresspool instance is using a stray ID fetch it by name.
+		found_addrpool = utils.GetSystemAddrPoolByName(addrpool_list, addrpool_instance.Name)
+
+	} else {
```
I don't think we need this else or the found_addrpool = utils.GetSystemAddrPoolByName(addrpool_list, addrpool_instance.Name) 2x. If found with UUID return, if not return using name.

sriram-gn replied:
It's updated now, please check

hugonicodemos commented on controllers/host/platformnetworks.go:
```diff
@@ -5,72 +5,911 @@ package host
+		// In case network instance is using a stray ID fetch it by name.
+		found_network = utils.GetSystemNetworkByName(network_list, network_instance.Name)
+
+	} else {
```
ditto about else, and repetition
